src/puppetlabs/trapperkeeper/services/webserver/jetty9_websockets.clj

```diff
@@ -0,0 +1,105 @@
+(ns puppetlabs.trapperkeeper.services.webserver.jetty9-websockets
+  (:import (clojure.lang IFn)
+           (org.eclipse.jetty.server Request)
+           (org.eclipse.jetty.websocket.api WebSocketAdapter Session)
+           (org.eclipse.jetty.websocket.server WebSocketHandler)
+           (org.eclipse.jetty.websocket.servlet WebSocketServletFactory WebSocketCreator)
+           (java.security.cert X509Certificate)
+           (java.nio ByteBuffer))
+
+  (:require [clojure.tools.logging :as log]
+            [puppetlabs.websockets.client :refer [WebSocketProtocol]]
+            [schema.core :as schema]))
+
+(def WebsocketHandlers
+  {(schema/optional-key :on-connect) IFn
+   (schema/optional-key :on-error) IFn
+   (schema/optional-key :on-close) IFn
+   (schema/optional-key :on-text) IFn
+   (schema/optional-key :on-bytes) IFn})
+
+(defprotocol WebSocketSend
+  (-send! [x ws] "How to encode content sent to the WebSocket clients"))
+
+(extend-protocol WebSocketSend
+  (Class/forName "[B")
+  (-send! [ba ws]
+    (-send! (ByteBuffer/wrap ba) ws))
+
+  ByteBuffer
+  (-send! [bb ws]
+    (-> ^WebSocketAdapter ws .getRemote (.sendBytes ^ByteBuffer bb)))
+
+  String
+  (-send! [s ws]
+    (-> ^WebSocketAdapter ws .getRemote (.sendString ^String s))))
+
+(extend-protocol WebSocketProtocol
+  WebSocketAdapter
+  (send! [this msg]
+    (-send! msg this))
+  (close! [this]
+    (.. this (getSession) (close)))
+  (remote-addr [this]
+    (.. this (getSession) (getRemoteAddress)))
+  (ssl? [this]
+    (.. this (getSession) (getUpgradeRequest) (isSecure)))
+  (peer-certs [this]
+    (.. this (getCerts)))
+  (idle-timeout! [this ms]
+    (.. this (getSession) (setIdleTimeout ^long ms)))
+  (connected? [this]
+    (. this (isConnected))))
+
+(defn- do-nothing [& args])
```

Comment:
Would it maybe be worth extending this so that it logged a message, at debug or trace level, indicating that an event had been handled by the default no-op handler?  Maybe change it to:

```clj
(defn- do-nothing [event & args]
   (log/debugf "No handler defined for websocket event '%s' with args: '%s'" event args))
```

and then close over it with `(partial do-nothing :on-connect)`, `(partial do-nothing :on-error)`, etc.?

Not 100% sure whether this is necessary, just thinking that maybe it might be useful for debugging purposes?